src/main/scala/dibl/NewPairDiagram.scala

```diff
@@ -0,0 +1,147 @@
+      seqNr += 1
+      SimpleNode(seqNr, point)
+    }
+
+    def toNodeSeq(row: Int, col: Int): Seq[ConnectedNode] = {
+      val item = itemMatrix(row)(col)
+
+      val relativeSources = toRelativeSources(item.vectorCode)
+      if (relativeSources.isEmpty) return Seq.empty // vectorCode == '-' or invalid
+      val Array((leftRow, leftCol), (rightRow, rightCol)) = relativeSources
+
+      val sourceLeft = toPoint(row + leftRow, col + leftCol)
+      val sourceRight = toPoint(row + rightRow, col + rightCol)
+      val target = toPoint(row, col)
+
+      val colorName = Stitches.defaultColorName(item.stitch)
+      seqNr += 1
+      val node = Node(seqNr, target, sourceLeft, sourceRight, item.stitch, item.id, colorName)
+      (isFringe(sourceLeft), isFringe(sourceRight)) match {
+        case (false, false) => Seq(node)
+        case (true, false) => Seq(node, toSimpleNode(sourceLeft, node, isLeft = true))
+        case (false, true) => Seq(node, toSimpleNode(sourceRight, node, isLeft = false))
+        case (true, true) => Seq(node, toSimpleNode(sourceLeft, node, isLeft = true), toSimpleNode(sourceRight, node, isLeft = false))
+      }
+    }
+
+    val nodes: Seq[ConnectedNode] = Seq(SimpleNode(0, Point(0, 0))) ++
+      (0 until cols).flatMap(row =>
+        (0 until rows).map(col =>
+          toNodeSeq(row, col)
+        )
+      ).flatten
+
+    // lookup table
+    val nodeMap: Map[Point, Int] = nodes.map(n => n.target -> n.seqNr).toMap
+
+    def findNode(point: Point): ConnectedNode = nodes(nodeMap.getOrElse(point, 0))
+
+    def toPairLinks(target: Node) = {
+      val leftNode = findNode(target.srcLeft)
+      val rightNode = findNode(target.srcRight)
+      Seq(
+        LinkProps.pairLink(
+          source = leftNode.seqNr,
+          target = target.seqNr,
+          start = leftNode.color,
+          mid = leftNode.twistsToLeftOf(target) - 1,
+          end = target.color),
+        LinkProps.pairLink(
+          source = rightNode.seqNr,
+          target = target.seqNr,
+          start = rightNode.color,
+          mid = rightNode.twistsToRightOf(target) - 1,
+          end = target.color)
+      )
+    }
+
+    var pairNr = 0
+    Diagram(
+      nodes.map {
+        case SimpleNode(_, Point(x, y)) =>
+          pairNr += 1 // TODO why does it start with 3?
+          NodeProps.node(s"Pair $pairNr", x, y)
+        case Node(_, Point(x, y), _, _, stitch, id, color) => NodeProps.node(s"$stitch - $id", color, x, y)
+      },
+      nodes.withFilter(_.isInstanceOf[Node])
+        .flatMap { case target: Node => toPairLinks(target) }
+    )
+  }
+
+  private trait ConnectedNode {
+    val seqNr: Int
+    val target: Point
+    val color: String
+
+    def twistsToLeftOf(target: Node) = 0
+
+    def twistsToRightOf(target: Node) = 0
+  }
+
+  private case class SimpleNode(override val seqNr: Int,
+                                override val target: Point
```

Comment:
Its confusing to call _the_ point of a simple node target. This was done to allow a super class and avoid having two cases when creating the nodeMap. Now we need the target to make make _the_ point of the SimpleNode unique when we search for it.